numalogic/blocks/pipeline.py

```diff
@@ -0,0 +1,173 @@
+        err_msg = (
+            "Invalid kwarg: {pname} found. Keyword args of "
+            "BlockPipeline must be of the form blockname__parameter, "
+            "e.g. `block_pipeline.fit(data, nn__max_epochs=50)`"
+        )
+        for pname, pval in fit_params.items():
+            if "__" not in pname:
+                raise ValueError(err_msg.format(pname=pname))
+            blockname, param = pname.split("__", 1)
+            block_params[blockname][param] = pval
+        return block_params
+
+    def fit(self, input_: npt.NDArray[float], **fit_params) -> npt.NDArray[float]:
+        """
+        Fit the pipeline on the input data.
+
+        Args:
+        ----
+            input_: The input data to fit the pipeline on.
+            fit_params : dict of string -> object
+            Parameters passed to the ``fit`` method of each block, where
+            each parameter name is prefixed such that parameter ``p`` for step
+            ``s`` has key ``s__p``.
+
+        Returns
+        -------
+            Final fit block output.
+        """
+        fit_params = self._get_block_params(**fit_params)
+        for block in self._blocks:
+            input_ = block.fit(input_, **fit_params.get(block.name, {}))
+        return input_
+
+    def run(self, data: npt.NDArray[float]) -> npt.NDArray[float]:
+        """
+        Perform inference on streaming data.
+
+        Args:
+        ----
+            data: Streaming input data
+
+        """
+        for block in self._blocks:
+            data = block.run(data)
+        return data
+
+    def save(self, skeys: Sequence[str], dkeys: Sequence[str]) -> None:
+        """
+        Save the state of the pipeline.
+
+        Args:
+        ----
+            skeys: Sequence of source keys.
+            dkeys: Sequence of destination keys.
+
+        Raises
+        ------
+            ValueError: If no registry is provided.
+        """
+        if not self._registry:
+            raise ValueError("No registry provided.")
+
+        artifacts: dict[str, artifact_t] = {}
+        for block in self._blocks:
+            if not block.stateful:
+                continue
+            artifacts[block.name] = block.artifact_state
+        self._registry.save(skeys, dkeys, artifacts)
+
+    def load(self, skeys: Sequence[str], dkeys: Sequence[str]) -> None:
+        """
+        Load the state of the pipeline.
+
+        Args:
+        ----
+            skeys: Sequence of source keys.
+            dkeys: Sequence of destination keys.
+
+        Raises
+        ------
+            ValueError: If no registry is provided.
+        """
+        if not self._registry:
+            raise ValueError("No registry provided.")
+
+        artifact_data = self._registry.load(skeys, dkeys)
+        artifacts = artifact_data.artifact
+        for block in self._blocks:
+            if not block.stateful:
+                continue
+            block.artifact_state = artifacts[block.name]
```

Comment:
Like the loading artifact mechanism!

We may have to be on a lookout of dict artifact that we are storing.

Reply:
Let's create an issue for this, because I think at this moment only redis registry will be able to support this.